### DISCUSSION TITLE
Unexpected failure of "CompileAvaloniaXamlTask" while trying to use Dictionary.Values on ListBox w/ CompiledBinding

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#9156] [Q&A]
I prefer opening this discussion first just to check if i'm not getting skill issued, this can be moved to issues if deemed necessary

repo: https://github.com/Mrcubix/Avalonia-ListBox-Issue

Version: 11.0.0-preview1 & 11.0.0-preview2

I'm trying to use the values of a provided dictionary to feed a ListBix, here containing a textbox
While using `CompileBindings=true`, the following compile error appear:

```
.nuget\packages\avalonia\11.0.0-preview2\buildTransitive\AvaloniaBuildTasks.targets(98,5): error MSB4018: 

Unexpected failure of task "CompileAvaloniaXamlTask". 

.nuget\packages\avalonia\11.0.0-preview2\buildTransitive\AvaloniaBuildTasks.targets(98,5): error MSB4018: 

System.ArgumentException: Member 'System.Collections.Generic.Dictionary`2/ValueCollection' is declared in another module and needs to be imported 
```

I don't think that's normal and i didn't see anything about it in [CompiledBindings's doc page](https://docs.avaloniaui.net/docs/data-binding/compiledbindings)

any ideas?

## Update

After reading a little more, it seems that i went past this comment that didn't seem useful at first

`All child nodes will inherit this property, so you can enable it in your root node and disable it for a specific child, if needed.`

So that mean i would have to set `x:DataType:ValueCollection` & import the namespace for the ListBox?

EDIT: nope apparently, not some related namespace or type either

**Comment by timunie** [comment]
I think I had the same issue back then. I ended up adding a new property in my `ViewModel` of type `IEnumerable` and just return `MyDict.Values` .

I'm not sure if that should be supported or not. Iirc the comes from the generic type of `Dictionary` 

If it's a bug, it should be fixed. Otherwise it should at least be documented.

**Reply by Mrcubix** [reply]
It work with ComboBoxes, so either remove support from combobox, or add it to every other controls supporting IEnumerable as Items

In the end it might not be a bug, but but it seem like a lack of consistency between controls.
(I believe it happen in ItemsControl too, tho i might be wrong)

Edit: Note again that this only happen when using compiled bindings, so it might already be the case, just for some reasons compiled Bindings break that

**Reply by timunie** [reply]
That doesn't make any sense, but let me check your sample

**Reply by timunie** [reply]
Upd: 

- It only fails if CompiledBindings are activated
- In that case it also fails for other ItemControls like ComboBox

possible Solution: 
Cast your Binding to a non-generic base-class or interface. I did that for you: https://github.com/Mrcubix/Avalonia-ListBox-Issue/pull/1 

```diff
+ xmlns:sys="using:System.Collections"

-    <ListBox Items="{Binding SomeDictionary.Values}" Width="395" Height="279">
+    <ListBox Items="{Binding ((sys:IDictionary)SomeDictionary).Values}" Width="395" Height="279">
```

**Reply by maxkatz6** [reply]
Fixed in master
